src/credentials/attestation_verifier/DefaultDeviceAttestationVerifier.cpp

```diff
@@ -636,13 +636,28 @@ AttestationVerificationResult DefaultDACVerifier::ValidateCertificationDeclarati
             ChipLogError(NotSpecified, "Disallowing CD signed by test key");
             return AttestationVerificationResult::kCertificationDeclarationNoCertificateFound;
         }
-
         ChipLogProgress(NotSpecified, "Allowing CD signed by test key");
     }
 
     VerifyOrReturnError(CMS_Verify(cmsEnvelopeBuffer, verifyingKey, certDeclBuffer) == CHIP_NO_ERROR,
                         AttestationVerificationResult::kCertificationDeclarationInvalidSignature);
 
+    // certDeclBuffer is populated by CMS_Verify so we need to do this check after the signature check
+    CertificationElementsWithoutPIDs cdContent;
+    VerifyOrReturnError(DecodeCertificationElements(certDeclBuffer, cdContent) == CHIP_NO_ERROR,
+                        AttestationVerificationResult::kCertificationDeclarationInvalidFormat);
+    // Ensure we didn't use a test key for official or provisional certificates (if disallowed)
+    bool testKeyAllowedForCertificateType =
```

Comment:
```suggestion
    bool testKeyAllowedForCertificationType =
```

Technically, the Certification Elements are not a certificate, yes?

Reply:
yup - totally. mistyped the field name. Will have to do this as a separate commit to get all the bits.